### PR TITLE
feat: add apiKeyRepairEnabled AppSync configuration option

### PR DESCRIPTION
**Pull request by bsantare**
Set apiKeyRepairEnabled to `true` to automatically create a new, unique AppSync API key CFN object if the plugin-maintained CFN object is associated with an AppSync API key id that no longer exists.  This can happen on API key exipration or if an API key is manually deleted from the AWS console.  If this condition is present, deployment will fail without this option enabled because CFN will try to look up the deleted API key by ID and fail.  This option is only relevant for authenticationType === API_KEY

* Modify mergeCustomProviderResources hook to be async
* Modify addResources to allow async calls
* Modify getApiKeyResources to handle key repairs when enabled
* Modify getApiKeyOutputs to use the resolvedLogicalApiKeyId resolved in getApiKeyResources

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -16,6 +16,9 @@ const RESOURCE_URL = 'GraphQlApiUrl';
 const RESOURCE_API_ID = 'GraphQlApiId';
 const RESOURCE_CACHING = 'GraphQlCaching';
 
+const PHYS_RES_ID_REGEX = new RegExp('^arn:aws:appsync:(?:.*:apis/)(?<apiId>.*)/apikeys/(?<apiKeyId>.*)$');
+const APPSYNC_API_KEY_ITER_REGEX = new RegExp('^GraphQlApiKeyDefault(?:_?)(?<apiKeyIter>.*)$');
```
Comment (jogura) [inline src/index.js:20]: is `GraphQlApiKeyDefault` defined as a constant somewhere? if so can it be referenced here?

File: src/index.js [hunk 2]
```diff
@@ -182,6 +186,56 @@ class ServerlessAppsyncPlugin {
       });
   }
 
+  listStackResources() {
```
Comment (jogura) [inline src/index.js:189]: Why wasn't this function declared async? Then we wouldn't need the `.then` at the end right?

File: src/index.js [hunk 3]
```diff
@@ -286,13 +340,15 @@ class ServerlessAppsyncPlugin {
...
 
-  addResources() {
+  async addResources() {
     const config = this.loadConfig();
 
     const resources = this.serverless.service.provider.compiledCloudFormationTemplate.Resources;
     const outputs = this.serverless.service.provider.compiledCloudFormationTemplate.Outputs;
 
-    config.forEach((apiConfig) => {
+    for (let i = 0; i < config.length; i += 1) {
```
Comment (jogura) [inline src/index.js:349]: consider adding a note to why this was changed from a `forEach` to a `for-loop`. I still get tripped up sometimes scratching my head why my async functions don't work.

File: src/index.js [hunk 4]
```diff
@@ -422,17 +479,128 @@ class ServerlessAppsyncPlugin {
...
+    }
+    return undefined;
+  }
+
+  /**
+   * Resolve the first matching API Key Resource for this API - there should only ever be 0 or 1
+   */
+  async resolveExistingApiKeyInfo({ baseKeyResourceIdName, appsyncApiName }) {
+    // Find the AppSync API id for the well-known API name
+    const appsyncApiId = await this.resolveAppSyncApiId(appsyncApiName);
```
Comment (jogura) [inline src/index.js:500]: Consider short circuiting here - if `resolveAppSyncApiId` returns an `undefined` the rest of the logic shouldn't matter right.

**Review by jogura — COMMENTED** [review]
Very clean & easy to follow. I left a few comments to consider.

File: src/index.js (hunk 1 above)
Comment (bsantare) [reply src/index.js]: Yep - changed

File: src/index.js (hunk 2 above)
Comment (bsantare) [reply src/index.js]: Yep - changed

File: src/index.js (hunk 4 above)
Comment (bsantare) [reply src/index.js:500]: Done

**Review by jogura — APPROVED** [review]
LGTM

**Pull request merged**